server/swap/swap.go

```diff
@@ -894,10 +1069,57 @@ func (s *Swapper) handleInit(user account.AccountID, msg *msgjson.Message) *msgj
 		return rpcErr
 	}
 
-	// Since we have to consider latency, run this as a coin waiter.
-	s.coinWaiter.Wait(coinwaiter.NewSettings(user, msg, params.CoinID, txWaitExpiration), func() bool {
-		return s.processInit(msg, params, stepInfo)
-	})
+	// init requests should only be sent when contracts are still required, in
+	// the correct sequence, and by the correct party.
+	switch stepInfo.match.Status {
+	case order.NewlyMatched:
+		if !stepInfo.actor.isMaker {
+			// s.step should have returned an error of code SettlementSequenceError.
+			panic("handleInit: this stepInformation should be for the maker!")
+		}
+	case order.MakerSwapCast:
+		if stepInfo.actor.isMaker {
+			// s.step should have returned an error of code SettlementSequenceError.
+			panic("handleInit: this stepInformation should be for the taker!")
+		}
+	default:
+		fmt.Println("Swap status:", stepInfo.match.Status)
+		return &msgjson.Error{
+			Code:    msgjson.SettlementSequenceError,
+			Message: "swap contract already provided",
+		}
+	}
+
+	// Validate the coinID and contract script before starting a coinWaiter.
+	err = stepInfo.asset.Backend.ValidateCoinID(params.CoinID)
+	if err != nil {
+		// TODO: ensure Backends provide sanitized errors or type information to
+		// provide more details to the client.
+		return &msgjson.Error{
+			Code:    msgjson.ContractError,
+			Message: "invalid contract coinID or script",
+		}
+	}
```

Comment:
The problem with `processInit`'s check is that it returns `coinwaiter.TryAgain` if the coinID decoding fails in `Coin`.  There is no distinction between that kind of error and simply not yet seeing the txn on the network yet.

```go
coin, err := chain.Coin(params.CoinID, params.Contract)
if err != nil {
	// If there is an error, don't give up yet, since it could be due to network
	// latency. Check again on the next tick.
...
	return coinwaiter.TryAgain
}
```

The idea of these initial checks is to avoid this retry loop that only ends on waiter expiration.

Reply:
Oh, regarding the initial check prior to creating the waiter:

https://github.com/chappjc/dcrdex/blob/f3d4596788020adc46ffe21e43a67ee97fd95504/server/coinwaiter/coinwaiter.go#L83-L86

```go
	// Check to see if it passes right away.
	if f() {
		return
	}
```

`f()` there returning `true`/`DontTryAgain` won't happen for a bad coinID as far as I can tell because a `Coin()` error returns `false`/`TryAgain`.

Reply:
I see now. Thanks for explaining. I was missing the `TryAgain` part.